python/src/object_store.rs

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use async_trait::async_trait;
+use chrono::{DateTime, Utc};
+use object_store::aws::AwsCredential;
+use object_store::CredentialProvider;
+use reqwest::Client;
+use serde::Deserialize;
+use std::fmt::{Debug, Formatter};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+use tokio::sync::Mutex;
+
+pub struct UrlBasedCredentialProvider {
+    url: String,
+    client: Client,
+    lock: Mutex<()>,
+    state: Mutex<CredentialState>,
+}
+
+#[derive(Deserialize, Debug)]
+struct CredentialState {
+    #[serde(rename = "AccessKeyId")]
+    access_key_id: Option<String>,
+
+    #[serde(rename = "SecretAccessKey")]
+    secret_access_key: Option<String>,
+
+    #[serde(rename = "SessionToken")]
+    session_token: Option<String>,
+
+    #[serde(rename = "ExpiredTime")]
+    expired_time: Option<SystemTime>,
+}
+
+impl UrlBasedCredentialProvider {
+    pub fn new(url: String) -> Self {
+        Self {
+            url,
+            client: Client::new(),
+            lock: Mutex::new(()),
+            state: Mutex::new(CredentialState {
+                access_key_id: None,
+                secret_access_key: None,
+                session_token: None,
+                expired_time: None,
+            }),
+        }
+    }
+
+    async fn try_get_credentials(&self) -> Option<AwsCredential> {
+        let state = self.state.lock().await;
+        if let Some(expiration) = state.expired_time {
+            if SystemTime::now() < expiration - Duration::from_secs(600) {
+                return Some(AwsCredential {
+                    key_id: state.access_key_id.clone().unwrap_or_default(),
+                    secret_key: state.secret_access_key.clone().unwrap_or_default(),
+                    token: state.session_token.clone(),
+                });
+            }
+        }
+        None
+    }
+}
+
+impl Debug for UrlBasedCredentialProvider {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "UrlBasedCredentialProvider {{ url: {} }}", self.url)
+    }
+}
+
+#[async_trait]
+impl CredentialProvider for UrlBasedCredentialProvider {
```

Comment:
Hi @eddyxu May I ask a question? I realized an `UrlBasedCredentialProvider ` for our inner object store service to get a temporal credential via an `assume role`. It seems it is very similar to the `AwsCredentialAdapter` you have provided. Based on `AwsCredentialAdapter`, if I only need to implement the `ProvideCredentials` trait, that's enough for my purpose?

Reply:
Is there any way you can re-use the `AwsCredentialsAdapter`? Or is the purpose of this one somehow different?

Reply:
Actually, I need to implement a non-standard `AssumeRoleProvider` which receives a URL and fetches `ak`, `sk` and `session_token` from it and support TTL. The built-in `AssumeRoleProvider` can not customize a URL. It seems AwsCredentialAdapter provided a TTL-based credential wrapper.  Maybe I need to implement `ProvideCredentials` trait.